src/phoenix-adapter.js

```diff
@@ -0,0 +1,258 @@
+    n = n + 1;
+    average = offsets.reduce((acc, offset) => (acc += offset), 0) / offsets.length;
+    if (offsets.length == numOffsetsToGather) {
+      setTimeout(update, 5 * 60 * 1000);
+    } else {
+      update();
+    }
+  };
+
+  update();
+
+  return () => {
+    return average;
+  };
+})();
+
+export default class PhoenixAdapter {
+  constructor() {
+    this.refs = new Map();
+    // TODO: Frozen messages can be handled outside this class.
+    this.frozenUpdates = new Map();
+    this._blockedClients = new Map();
+  }
+  setServerUrl() {}
+  setApp() {}
+  setRoom() {}
+  setWebRtcOptions() {}
+  setServerConnectListeners(nafConnectSuccess, nafConnectFailed) {
+    this.nafConnectSuccess = nafConnectSuccess;
+    this.nafConnectFailed = nafConnectFailed;
+  }
+  setRoomOccupantListener() {}
+  setDataChannelListeners(nafOccupantJoined, nafOccupantLeave, nafMessageReceived) {
+    this.nafOccupantJoined = nafOccupantJoined;
+    this.nafOccupantLeave = nafOccupantLeave;
+    this.nafMessageReceived = nafMessageReceived;
+  }
+  async connect() {
+    this.refs.set("naf", this.hubChannel.channel.on("naf", this.handleIncomingNAF));
+    this.refs.set("nafr", this.hubChannel.channel.on("nafr", this.handleIncomingNAFR));
+
+    this.nafConnectSuccess(this.session_id);
+    this.reliableTransport = transportForChannel(this.hubChannel.channel, true);
+    this.unreliableTransport = transportForChannel(this.hubChannel.channel, false);
+
+    this.hubChannel.presence.list(key => key).forEach(this.nafOccupantJoined);
+    this.refs.set("hub:join", this.events.on(`hub:join`, ({ key }) => this.nafOccupantJoined(key)));
+    this.refs.set("hub:leave", this.events.on(`hub:leave`, ({ key }) => this.nafOccupantLeave(key)));
+  }
+  shouldStartConnectionTo() {}
+  startStreamConnection() {}
+  closeStreamConnection() {}
+  getConnectStatus() {}
+
+  getMediaStream() {
+    return Promise.reject("getMediaStream not implemented in phoenix-adapter");
+  }
+
+  getServerTime() {
+    return Date.now() + getAverageTimeOffset();
+  }
+
+  sendData(clientId, dataType, data) {
+    this.unreliableTransport(clientId, dataType, data);
+  }
+  sendDataGuaranteed(clientId, dataType, data) {
+    this.reliableTransport(clientId, dataType, data);
+  }
+  broadcastData(dataType, data) {
+    this.unreliableTransport(undefined, dataType, data);
+  }
+  broadcastDataGuaranteed(dataType, data) {
+    this.reliableTransport(undefined, dataType, data);
+  }
+
+  disconnect() {
+    this.hubChannel.presence.list(key => key).forEach(this.nafOccupantLeave);
+
+    this.events.off("naf", this.refs.get("naf"));
+    this.events.off("nafr", this.refs.get("nafr"));
+    this.events.off("hub:join", this.refs.get("hub:join"));
+    this.events.off("hub:leave", this.refs.get("hub:leave"));
+    this.refs.delete("naf");
+    this.refs.delete("nafr");
+    this.refs.delete("hub:join");
+    this.refs.delete("hub:leave");
+  }
+
+  toggleFreeze() {
```

Comment:
All of the code about freezing, unfreezing, flushing/storing/reducing updates is all taken from `naf-dialog-adapter`.